### ISSUE TITLE
Appium IOS WebDriverAgent on IOS Simulators 

### ISSUE THREAD

**Issue by busrabasbahcivan** [appium/appium#14605]
## The problem

Hello, I am team-mate with sergydimitrov whom you worked with on the issue #14473. 

After the change you did (https://github.com/appium/WebDriverAgent/pull/355/commits/faa16ecff46ac67f5825ca08a6b5c4c18942a646), the Web Driver Agent could not start on our PID=0 condition. Could you please roll back your code change by deleting below code block?

 if (0 == processID) {
    return nil;
  }

Otherwise we should manually delete this block, after installing the new version of Appium, then we should manually build Web Driver Agent.

Thank you very much.

**Comment by mykola-mokhnach** [comment]
Please provide the logs for both cases (with and without the condition) and (if possible) explain in more details the need for such change, so we could make a proper decision.

**Comment by busrabasbahcivan** [comment]
Hello @mykola-mokhnach,

"Without the if condition", the logs are on this link which sergydimitrov posted: https://gist.github.com/sergydimitrov/d082cdfbddcc8bf725809a85332f7f24

You know, you could not repro our issue on your environment. And you did a code change. After your change, sergydimitrov and I tested the case again, the web driver agent could not be installed on the simulator. Sorry for that, I cannot share the "with condition logs" because the logs are our customer' s product' s logs, they don' t allow us to share the logs.

In the end, we upgraded to Appium 1.18.0 Release and deleted the if condition, and built the Web Driver Agent manually. Then, we could repro our pID=0 error again and did a research then found a good solution. If we add "waitForQuiescence : false" and "wdaEventloopIdleDelay:  3" capabilities to Appium or to our source code, our application could start immediately despite Catalina verifying pop-up.

To sum up, we tested the if condition you added and it doesn' t work on our problem. Because of that, if anyone has our problem, they won' t know the solution (removing the if condition) like us. This may be problem in the future.

Thank you.

**Comment by mykola-mokhnach** [comment]
Sorry, but I'm still not quite convinced.

This is the code, that is responsible for active app selection:

```
  if (nil != activeApplicationElement) {
    FBApplication *application = [FBApplication fb_applicationWithPID:activeApplicationElement.processIdentifier];
    if (nil != application) {
      return application;
    }
    [FBLogger log:@"Cannot translate the active process identifier into an application object"];
  }

  if (activeApplicationElements.count > 0) {
    [FBLogger logFmt:@"Getting the most recent active application (out of %@ total items)", @(activeApplicationElements.count)];
    for (XCAccessibilityElement *appElement in activeApplicationElements) {
      FBApplication *application = [FBApplication fb_applicationWithPID:appElement.processIdentifier];
      if (nil != application) {
        return application;
      }
    }
  }

  [FBLogger log:@"Cannot retrieve any active applications. Assuming the system application is the active one"];
  return [self fb_systemApplication];
```

without the condition `fb_applicationWithPID` would throw an exception and make the API call fail completely. Otherwise a system application would be returned, which is correct, because it would be active otherwise.

**Comment by mykola-mokhnach** [comment]
@KazuCocoa Would you like to add something on that?

**Comment by busrabasbahcivan** [comment]
> Sorry, but I'm still not quite convinced.
> 
> This is the code, that is responsible for active app selection:
> 
> ```
>   if (nil != activeApplicationElement) {
>     FBApplication *application = [FBApplication fb_applicationWithPID:activeApplicationElement.processIdentifier];
>     if (nil != application) {
>       return application;
>     }
>     [FBLogger log:@"Cannot translate the active process identifier into an application object"];
>   }
> 
>   if (activeApplicationElements.count > 0) {
>     [FBLogger logFmt:@"Getting the most recent active application (out of %@ total items)", @(activeApplicationElements.count)];
>     for (XCAccessibilityElement *appElement in activeApplicationElements) {
>       FBApplication *application = [FBApplication fb_applicationWithPID:appElement.processIdentifier];
>       if (nil != application) {
>         return application;
>       }
>     }
>   }
> 
>   [FBLogger log:@"Cannot retrieve any active applications. Assuming the system application is the active one"];
>   return [self fb_systemApplication];
> ```
> 
> without the condition `fb_applicationWithPID` would throw an exception and make the API call fail completely. Otherwise a system application would be returned, which is correct, because it would be active otherwise.

Hello again @mykola-mokhnach,

Without the condition, our problem was the verification popup. It was failing to create the WDA session because it sees Catalina Scheduler process id 0. Second try appium was creating the WDA session but it was too late (about 6 minutes) then test were stopped proceeding. All the detailed info is on the issue #14473 

With the condition, I think our situation runs your added "if condition", then it returns nil. Somehow, the WebDriverAgent application couldn' t be installed to the simulator. We tested it again and again and we are sure that your change doesn' t work, and it can be a problem in the future.

**Comment by mykola-mokhnach** [comment]
> With the condition, I think our situation runs your added "if condition", then it returns nil. Somehow, the WebDriverAgent application couldn' t be installed to the simulator.

Now I'm really confused. How it is possible that changing this one condition would prevent the whole app to be installed? 
What is written to the Simulator log when it fails? 

How it is possible the stuff works properly in CII and on my machine and on other contributor's machines?

**Comment by Dan-Maor** [comment]
I have to admit I'm a bit confused as well.

The original stacktrace showed that the `applicationProcessWithPID` API was called with pid 0, which is invalid, otherwise this exception would not have been thrown.

> In the end, we upgraded to Appium 1.18.0 Release and deleted the if condition, and built the Web Driver Agent manually. Then, we could repro our pID=0 error again

So if the condition is removed and the issue happened again afterwards, I believe that essentially the fix is correct.

This condition should not prevent WebDriverAgent from being installed though, unless the source code is not clean on the test machine. Perhaps try to completely uninstall and reinstall Appium from npm.

You mentioned that by using the `waitForQuiescence` and `wdaEventloopIdleDelay` capabilities you were able to resolve the launch issue, these were indeed added to help with cases where application launches would fail and I see that using them resolved the problem with your tests.

Will it be possible to share the Appium server log of the failed installation attempt while the condition is in place? Preferably after uninstalling and reinstalling Appium completely from the test machine, so that we'll start with a clean slate.

**Comment by busrabasbahcivan** [comment]
> > With the condition, I think our situation runs your added "if condition", then it returns nil. Somehow, the WebDriverAgent application couldn' t be installed to the simulator.
> 
> Now I'm really confused. How it is possible that changing this one condition would prevent the whole app to be installed?
> What is written to the Simulator log when it fails?
> 
> How it is possible the stuff works properly in CI and on my machine and on other contributor's machines?

**Comment by busrabasbahcivan** [comment]
> > With the condition, I think our situation runs your added "if condition", then it returns nil. Somehow, the WebDriverAgent application couldn' t be installed to the simulator.
> 
> Now I'm really confused. How it is possible that changing this one condition would prevent the whole app to be installed?
> What is written to the Simulator log when it fails?
> 
> How it is possible the stuff works properly in CI and on my machine and on other contributor's machines?

Hi again,

You didn' t encounter the problem which we got on our customer' s environment. So, your situation didn' t run the "if condition block".

**Comment by busrabasbahcivan** [comment]
Hi again @mykola-mokhnach and @Dan-Maor,

Okey, I understood your detailed explanation. Thank you very much for sharing your kindly opinions. I will try again on our customer' s machine and share the logs/results with you with their permisson. Then, we may examine the problem again more clearly.

Thank you again.
